docs/coding-guidelines.md

```diff
@@ -0,0 +1,402 @@
+
+Note: Public means callable by a customer, so it includes protected APIs. However, some public APIs might still be "for internal use only" but need to be public for technical reasons. We will still have doc comments for these APIs but they will be documented as appropriate.
+
+### Assertions
+
+Do not use `Debug.Assert()`. That's what unit tests are for.
+Consider using `Assumes.Present()`, specifically in Visual Studio code when interacting with the service providers.
+
+### Unit tests and functional tests
+
+#### Assembly naming
+
+The unit tests for the `NuGet.Fruit` assembly live in the `NuGet.Fruit.Tests` assembly.
+
+The functional tests for the `NuGet.Fruit` assembly live in the `NuGet.Fruit.FunctionalTests` assembly.
+
+In general there should be exactly one unit test assembly for each product runtime assembly. In general there should be one functional test assembly per product (NuGet.exe/MSBuild.exe/dotnet.exe). Exceptions can be made for both. Some are already grandfathered in.
+
+#### Unit test class naming
+
+Test class names end with `Test` and live in a similar namespace as the class being tested. For example, the unit tests for the `NuGet.Fruit.Banana` class would be in a `NuGet.Fruit.Test.BananaTest` class in the test assembly.
+
+#### Unit test method naming
+
+Unit test method names must be descriptive about *what is being tested*, *under what conditions*, and *what the expectations are*. Pascal casing and underscores can be used to improve readability. The following test names are correct:
+
+```cs
+PublicApiArgumentsShouldNotBeNull
+Public_api_arguments_should_not_be_null
+```
+
+The following test names are incorrect:
+
+```cs
+Test1
+Constructor
+FormatString
+GetData
+```
+
+#### Unit test structure
+
+The contents of every unit test should be split into three distinct stages, optionally separated by these comments:
+
+```cs
+// Arrange  
+// Act  
+// Assert
+```
+
+The crucial thing here is that the `Act` stage is exactly one statement. That one statement is nothing more than a call to the one method that you are trying to test. Keeping that one statement as simple as possible is also very important. For example, this is not ideal:
+
+```cs
+int result = myObj.CallSomeMethod(GetComplexParam1(), GetComplexParam2(), GetComplexParam3());
+```
+
+This style is not recommended because way too many things can go wrong in this one statement. All the `GetComplexParamN()` calls can throw for a variety of reasons unrelated to the test itself. It is thus unclear to someone running into a problem why the failure occurred.
+
+The ideal pattern is to move the complex parameter building into the `Arrange` section:
+
+```cs
+// Arrange
+P1 p1 = GetComplexParam1();
+P2 p2 = GetComplexParam2();
+P3 p3 = GetComplexParam3();
+
+// Act
+int result = myObj.CallSomeMethod(p1, p2, p3);
+
+// Assert
+Assert.AreEqual(1234, result);
+```
+
```

Comment:
Is it possible to add a convention that each unit test should test only one part? i.e., don't make a giant unit test that test multiple parts of your code

Reply:
What do you have in mind? 

Does this cover it?
```
The crucial thing here is that the `Act` stage is exactly one statement. That one statement is nothing more than a call to the one method that you are trying to test. Keeping that one statement as simple as possible is also very important. 
```